src/s3/util/fileViewerManager.ts

```diff
@@ -2,30 +2,198 @@
+        this.window.showTextDocument(fileLocation)
+    }
+
+    /**
+     * Fetches a file from S3 or gets it from the local cache if possible.
+     */
+    public async getFile(fileNode: S3FileNode): Promise<vscode.Uri | undefined> {
+        if (!this.tempLocation) {
+            await this.createTemp()
+        }
+        const targetPath = await this.createTargetPath(fileNode)
+        const targetLocation = vscode.Uri.file(targetPath)
+
+        const tempFile: vscode.Uri | undefined = await this.getFromTemp(fileNode)
+        //if it was found in temp, return the Uri location
+        if (tempFile) {
+            return tempFile
+        }
+
+        //needs to be downloaded from S3
+        //if file is +4MB, prompt user to confirm before proceeding
+        //const uri = vscode.Uri.file('')
+
+        if (fileNode.file.sizeBytes === undefined) {
+            const message = "File size couldn't be determined. Continue with download?"
+            if (!(await this.promptUserConfirm(message))) return undefined
+        } else if (fileNode.file.sizeBytes > SIZE_LIMIT) {
+            showOutputMessage(`size is >4MB, prompt user working`, this.outputChannel)
+
+            const message = 'File size is more than 4MB. Continue with download?'
+            if (!(await this.promptUserConfirm(message))) return undefined
+
+            showOutputMessage(`user confirmed download, continuing`, this.outputChannel) //TODOD:: debug log,
+        }
+
+        //good to continue with download
+        try {
+            await downloadWithProgress(fileNode, targetLocation, this.window)
+        } catch (err) {
+            showOutputMessage(`error calling downloadWithProgress: ${err.toString()}`, this.outputChannel)
+        }
+
+        this.cacheArns.add(fileNode.file.arn)
+        //await this.listTempFolder()
+
+        return targetLocation
     }
 
-    public async getFile() {}
+    async getFromTemp(fileNode: S3FileNode): Promise<vscode.Uri | undefined> {
+        const targetPath = await this.createTargetPath(fileNode)
+        const targetLocation = vscode.Uri.file(targetPath)
+
+        if (this.cacheArns.has(fileNode.file.arn)) {
+            //get it from temp IF it hasn't been recently modified, then return that
+            showOutputMessage(`cache is working!, found ${fileNode.file.key} in cache`, this.outputChannel) //TODOD:: debug log remove
+            //explorer (or at least the S3Node) needs to be refreshed to get the last modified date from S3
+            const newNode = await this.refreshNode(fileNode)
+            if (!newNode) {
+                showOutputMessage(
+                    `!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! something happeneded`,
+                    this.outputChannel
+                ) //TODOD:: debug log, remove
+                return
+            }
+
+            fileNode = newNode
+            const lastModifiedInS3 = fileNode!.file.lastModified
```

Comment:
We should always make a network call for the most recent lastModified date. If a user updated a file through other means and hasn't refreshed their S3 node in a while, this will continue to pull the cached file.

Reply:
That is why I have `refreshNode()` which gets the last updated instance of the S3FileNode